# Masters' in Applied Artificial Intelligence
## Machine Learning Algorithms Course

Notebooks for MLA course

by [*lufer*](mailto:lufer@ipca.pt)

(ver 3.0)

---

# Part I - Datasets for ML Modelling



*   This part explores the main considerations about datasets, namely: i) the way to get it; ii) main statistics involved; iii) essencial manipulations actions.
*   Main python topics and concepts are also explored.


**Contents**:

0. **Environment preparation**
1. **Python essencial**
2. **Pytho OOP**
3. **Understanding Datasets**
4. **References**




## 0 - Environment preparation


Verify if python is installed; check the drive for notebooks files;

In [1]:
#install phyton
#check if python is installed
!python --version # Python >= 3.9.18?

Python 3.12.12


### Import necessary Libraries

In [2]:
import pandas as pd
import numpy as np

### Mounting Google Drive

In [3]:
#mount my google onedrive
#from google.colab import drive
#drive.mount('/content/drive')

In [7]:
# or
#see https://towardsdatascience.com/different-ways-to-connect-google-drive-to-a-google-colab-notebook-pt-1-de03433d2f7a
#see https://adriandolinay.medium.com/an-introduction-to-google-colab-2023-6c26792827b3

from google.colab import drive

# it will ask for your google drive credentiaals
drive.mount('/content/MIAADrive/', force_remount=True)

Mounted at /content/MIAADrive/


### Unmounting Google Drive

In [5]:
from google.colab import drive
drive.mount('/content/MIAADrive')

Mounted at /content/drive


In [6]:
#see https://www.geeksforgeeks.org/unmount-drive-in-google-colab-and-remount-to-another-drive/
#!ls "/content/MIAADrive/"
#from google.colab import drive
#drive.flush_and_unmount()

## 1 - Python Essential


This section intends to recover main ython concepts to work in ML

### Python Functions Definition

In [ ]:
%pwd


In [ ]:
#phyton functions definition
def show(p): print (p)

#using defined functions
show("Lets start working with DataSets")

**Create an external libary of functions**


In [8]:
#Importing external python scripts (models)
%cd "/content/MIAADrive/MyDrive/Colab Notebooks/MIAA - ML - 2025-2026/Functions"
%pwd
#import
import PythonFunctions as my  #PythonFunction.py library
print(my.addNumbers(2,3))
print(my.getCurrentTime())
print(my.subNumbers(1,2))

#After any change in external module, reload it!
import importlib
importlib.reload(my)

/content/MIAADrive/MyDrive/Colab Notebooks/MIAA - ML - 2025-2026/Functions
5
2025-11-17 17:38:22
-1


<module 'PythonFunctions' from '/content/MIAADrive/MyDrive/Colab Notebooks/MIAA - ML - 2025-2026/Functions/PythonFunctions.py'>

**Documenting python code**

In each function define help documentation: Description; Parameters, Returns, Examples

```
def addNumbers(a, b):
    """
    Add two numbers together.

    Parameters
    ----------
    a : int or float
        The first number to add.
    b : int or float
        The second number to add.

    Returns
    -------
    int or float
        The sum of `a` and `b`.

    Examples
    --------
    >>> addNumbers(3, 5)
    8
    >>> addNumbers(2.5, 4.5)
    7.0
    """
    return a + b
```



**Get documentation about any function**


```
help(my.addNumbers)
#or
print(my.addNumbers.__doc__)
```



In [9]:
#to get help about any fucntion
#help(my.addNumbers)
#or
print(my.addNumbers.__doc__)


    Add two numbers together.

    Parameters
    ----------
    a : int or float
        The first number to add.
    b : int or float
        The second number to add.

    Returns
    -------
    int or float
        The sum of `a` and `b`.

    Examples
    --------
    >>> addNumbers(3, 5)
    8
    >>> addNumbers(2.5, 4.5)
    7.0
    


**Things about the operating system**

In [ ]:
#using module "os"
import os
print (os.getcwd())
os.listdir()
#os.rename()
#os.chdir(path)
#explore more functions
#see https://www.w3schools.com/python/module_os.asp

In [ ]:
#using module shutil for copying, renaming, etc.
import shutil
shutil.copy("PythonFunctions.py", "OtherCopied.py")
#shutil.move("OtherCopied.py","OtherCopiedII.py")

In [ ]:
os.listdir()

**Handling execution time**

In [ ]:
#Other example
import time #perf_counter(), time(), etc...see https://builtin.com/articles/timing-functions-python
print("Sleeping")
time.sleep(10) # sleep for a while; interrupt me!
#equivalent to
#for i in range(1,10):
#  time.sleep(1)
#  print(i)
print("Done Sleeping for 10 seconds")



In [ ]:
#using time to know the time elapsed on the execution...
ini=time.perf_counter()
my.getCurrentTime()
fim=time.perf_counter()
a = f'{(fim-ini):.3}'
print("Elapsed time: ",a,"s")

###More Python

**Lists**

In [ ]:
#Lists
#uncoment each instruction and analyse the result
l = [1,2,3,2,5,6]
#type(l) #list
#len(l)
#l[0]
#l[-2]
#l[-6]

#l[start_slice_value: end_slide_value]
#In a slice l[start:end]:
  #The start index is included.
  #The end index is excluded.

#l[2:4]  #from index 2 (inclusive) to 4 (exclusive)
#l[:]
#l[2:]
#l[:4]


**Series**

In [ ]:
#Series: like an array
s = pd.Series([13,23,4,123,-4,134])
#type(s)  #analyse the output
s

In [ ]:
#what happens?
s = pd.Series([13,23,4,123],index=['a','b','c','d'])
#s
#s[0]
#s[1:3]
#s[:]
#s[:-1]
#s[1:2]



**Dictionary**

In [ ]:
#HashTable or Dictionary
# { label: serie,...,label:serie}

data = {
    "Colors":["Red","White","Orange","Black"],
    "Paintings":["Monalisa","The Kiss", "The Last Supper","O GLorioso"]
     }
type(data)
data["Colors"]
data["Paintings"]
data["Paintings"][0]

In [ ]:
#and this case?
#data = {
#    "Colors":{"Red","White","Orange","Black"},
#    "Paintings":{"Monalisa","The Kiss", "The Last Supper","O GLorioso"}
#    }
#type(data)
#data["Colors"]
#data["Colors"][0}

**Dataframe**

In [ ]:
#Dataframes: two-dimensional array of series

df = pd.DataFrame(data)
type(df)
df.head()

## 2 - Python Object Oriented (POO)

Object programing means the existence of classes (models) and objects (instances)


### Classes definition

In [ ]:
#Class definition
#lufer
class Person:
  #-----
  #Attributes | state
  name=""
  age=1
  #-----
  #Methods | Behaviour
  # assign values to object properties, or other operations that are necessary
  # to do when the object is being created
  #Constructor
  def __init__(self, name, age):
    #print("Inside __init__")
    self.name = name
    self.age = age
  #-----
  #  controls what should be returned when the class object is represented as a string
  def __str__(self):
    return f"Name: {self.name} (Age: {self.age})"
  #-----
  # other methods
  def myName(self):
    print("Hello my name is  " + self.name)

  #class methods
  def myAge(self):
    print("I am " + str(self.age) + " years old")

### Objects creation and using

In [ ]:
#Objects
p1 = Person("Ana Gustavo","12")   #uses __init__

print(p1)                         #uses __str__
print(p1.age)
print (f"Age: ({p1.age}) - Name: ({p1.name}) " )
p1.myName();
p1.name

Name: Ana Gustavo (Age: 12)
12
Age: (12) - Name: (Ana Gustavo) 
Hello my name is  Ana Gustavo


'Ana Gustavo'

### Using Class Methods

In [ ]:
#class methods are used my class name
Person.myAge(p1)

I am 12 years old


### Python's .*\_\_call__()* Method [link text](https://)

*\_\_call()__* is an alternative to make "callable" methods or variables. Many actual python versions does not support this.

**NOTE:**
* Every Python class is callable, meaning that calling it creates a new instance of the class
* *\_\_call__* can generate complexity

In [ ]:
#callable(abs)
#check if it is callable
callable (Person.myAge)

True

In [ ]:
#Another  example
class PersonII:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def Hi(self):
        return f"Hi, I'm {self.name} and I'm {self.age} years old."

    def MyAge(self):
        return f"My age is {self.age}"

    #def __call__(self):
    #    # Calling the object will call MyAge()
    #    return self.MyAge()

    def __call__(self, action=None):
        if action is None:
            # Default callable behavior
            return self.MyAge()

        if action == "Hi":
            return self.Hi()

        return f"Unknown action: {action}"


In [ ]:
#using it

a=PersonII("ola",12)
a.Hi()

"Hi, I'm ola and I'm 12 years old."

In [ ]:
#using the callable method
a()

a("Hi")

"Hi, I'm ola and I'm 12 years old."

### Código HTML

In [ ]:
%%html
<marquee style='width: 30%; color: red;'><b>O Benfica é o maior!</b></marquee>

### HTML/SVG

In [ ]:
%%html
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 450 400" width="200" height="200">
  <rect x="80" y="60" width="250" height="250" rx="20" style="fill:red; stroke:black; fill-opacity:0.7" />
  <rect x="180" y="110" width="250" height="250" rx="40" style="fill:blue; stroke:black; fill-opacity:0.5;" />
</svg>

## 3 - Understanding Datasets

Datasets are sets of data to be used on models training. A dataset can have Continous values (quantities); Categorical values (fixed list values) or Text values.

### Read a Dataset from a CSV file

In [13]:
filePath="/content/MIAADrive/MyDrive/Colab Notebooks/MIAA - ML - 2025-2026/"

In [ ]:
#option 1
with open(filePath+'Datasets/test.csv', 'r') as f:
    print(f.read())

In [20]:
#option 2
dataset = pd.read_csv(filePath+'/Datasets/test.csv')
#dataset is a dataframe

In [21]:
dataset.head()

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   
2   3           1807     1          2.8         0   1       0          27   
3   4           1546     0          0.5         1  18       1          25   
4   5           1434     0          1.4         0  11       1          49   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   
1    0.8        191  ...  12        746       857  3895     6     0   
2    0.9        186  ...   4       1270      1366  2396    17    10   
3    0.5         96  ...  20        295      1752  3893    10     0   
4    0.5        108  ...  18        749       810  1773    15     8   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  
1          7        1             0     0  
2         10        0             1     1  
3          7        1             1     0  
4          7        1             0     1  

[5 rows x 21 columns]

In [ ]:
#type(dataset)
dataset.dtypes

After read, the dataset is a Dataframe python type. Pandas DataFrame is like an Excel-like table.

In [15]:
#option 3
# the "r" means ignoring special chars if the file name
f = open(filePath+r"/Datasets/test.csv", "r")
#f is a stream handler

In [16]:
type(f)

_io.TextIOWrapper

###Generate Datasets

* Python offers a simple and powerful way to generate data...*Faker* is one of the libraries;

* Read more about "Sintetic Datasets"


In [17]:
pip install faker

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 25.6 MB/s eta 0:00:00


In [19]:
from faker import Faker
import datetime
import random

# Set the seed for reproducibility
random.seed(42)

# Generate random sales data
fake = Faker()
sales_data = []

for _ in range(1000):
    city = fake.city()
    state = fake.state()
    sale = round(random.uniform(300, 5000), 2)
    sale_date = fake.date_between(start_date='-1y', end_date='today')
    sales_data.append([city, state, sale, sale_date])

# Create a DataFrame
df = pd.DataFrame(sales_data, columns=['City', 'State', 'Sale Amount', 'Sale Date'])

# Display the first few rows of the dataset
print(df.head())

                City         State  Sale Amount   Sale Date
0    New Robertmouth      Oklahoma      3305.31  2024-12-22
1           Foxville    Washington       417.55  2025-07-15
2        Rebeccafurt       Alabama      1592.64  2025-09-01
3  Lake Patriciastad          Ohio      1349.09  2025-06-13
4   East Zacharyland  North Dakota      3761.41  2024-11-29


In [ ]:
#or using a function

def generateSales(num_customers):
    customers = []
    for _ in range(num_customers):
        customer_id = fake.uuid4()
        name = fake.name()
        email = fake.email()
        customers.append({'CustomerID': customer_id, 'CustomerName': name, 'Email': email})
    return pd.DataFrame(customers)

df1 = generateSales(100)
df1.head()

###Save a Dataset to file

In [ ]:
# Save the dataset to a CSV file
df.to_csv(filePath+'/Datasets/sales_dataset.csv', index=False)

### Datasets and Dataframes



Working with datasets in ML involves many functions, like:

*len(), shape, axes, index, values, info(), head(), tail(), drop(), iloc()*

**Number of records in the dataframe**

In [ ]:
#number of records in the dataframe
len(dataset)

**Dataframe Structure (rows x columns)**

In [ ]:
#Structure (rows x columns) of the dataframe
dataset.shape

**Dataframe Axis (0 and 1)**

In [ ]:
dataset.axes
#RangeIndex = lines
#Index = columns

In [ ]:
#Columns
dataset.axes[1]

In [ ]:
#what does dataset.axes[0]?

In [ ]:
dataset.index

In [ ]:
dataset.values

**First 10 records**

In [ ]:
#first 10 records
dataset.head()

**Last 10 records**

In [ ]:
#last 10 records
dataset.tail()

**Details about each column of the dataset**

Count of *Data type* and *Non-Null* values *Texto em itálico*

In [ ]:
#Details about each column
dataset.info()

Indexing just the rows

In [ ]:
#first column
dataset.iloc[0]

In [ ]:
#first line (line 0)
dataset.iloc[[0]]

In [ ]:
dataset.iloc[[0, 1]]

With a slice object

In [ ]:
dataset.iloc[:3]

Indexing lines and columns (both indexes)

In [ ]:
dataset.iloc[0,1] #iloc[row,column]

In [ ]:
#lines from 1 to 3 and first 3 columns
dataset.iloc[1:3, 0:3]

In [ ]:
#all
dataset.iloc[:, :]

Droping data

In [ ]:
dataset.drop(['blue', 'four_g'], axis=1)
#or
#dataset.drop(columns=['blue, 'four_g'])
#or
#dataset.drop(['blue', 'four_g'], axis='columns')

In [ ]:
pd.set_option("display.precision", 2)


### Basic Statistics over a Dataset

The basic statisics of a dataset can be obtained using the python function *describe()*.

1. For Numeric Columns (default behavior), it provides summary statistics such as:
 * Count: The number of non-missing values.
 * Mean: The average of the values.
 * Std: The standard deviation (measure of spread).
 * Min: The smallest value.
 * 25%: The first quartile (25th percentile).
 * 50%: The median (50th percentile).
 * 75%: The third quartile (75th percentile).
 * Max: The largest value.
2.  For Non-Numeric Columns (if specified using include='all' or include=object), it provides:
 * Count: The number of non-missing values.
 * Unique: The number of unique values.
 * Top: The most frequent value.
 * Freq: The frequency of the top value.

3. **Key Notes about 'describe()'**
* Non-numeric data is ignored unless explicitly included using include.
* The method helps quickly identify:
 * Missing data (count vs. total rows).
 * Outliers (e.g., large differences between min, max, and mean).
 * Skewness (if the mean and median differ significantly).

 * **Default Percentiles**: percentiles=None computes the 25th,  50th, and 75th percentiles by default.
* **Default Inclusion**: exclude=None includes numeric columns only, unless explicitly overridden.
* Explicit Parameters:
 * Use percentiles to customize which percentiles to compute.
 * Use include='all' to analyze both numeric and non-numeric columns.
 * Use exclude to omit specific data types (e.g., exclude=[float]).

In [22]:
#Load the dataset
dataset = pd.read_csv(filePath+'/Datasets/test.csv')
#dataset is a dataframe

In [23]:
#see https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
#by default
dataset.describe()
#what happens with
#dataset.describe(include='all')
#How do you explain it?

id  battery_power         blue  clock_speed     dual_sim  \
count  1000.000000    1000.000000  1000.000000  1000.000000  1000.000000   
mean    500.500000    1248.510000     0.516000     1.540900     0.517000   
std     288.819436     432.458227     0.499994     0.829268     0.499961   
min       1.000000     500.000000     0.000000     0.500000     0.000000   
25%     250.750000     895.000000     0.000000     0.700000     0.000000   
50%     500.500000    1246.500000     1.000000     1.500000     1.000000   
75%     750.250000    1629.250000     1.000000     2.300000     1.000000   
max    1000.000000    1999.000000     1.000000     3.000000     1.000000   

                fc       four_g   int_memory        m_dep   mobile_wt  ...  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.00000  ...   
mean      4.593000     0.487000    33.652000     0.517500   139.51100  ...   
std       4.463325     0.500081    18.128694     0.280861    34.85155  ...   
min       0.000000     0.000000     2.000000     0.100000    80.00000  ...   
25%       1.000000     0.000000    18.000000     0.300000   109.75000  ...   
50%       3.000000     0.000000    34.500000     0.500000   139.00000  ...   
75%       7.000000     1.000000    49.000000     0.800000   170.00000  ...   
max      19.000000     1.000000    64.000000     1.000000   200.00000  ...   

                pc    px_height     px_width          ram         sc_h  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     10.054000   627.121000  1239.774000  2138.998000    11.995000   
std       6.095099   432.929699   439.670981  1088.092278     4.320607   
min       0.000000     0.000000   501.000000   263.000000     5.000000   
25%       5.000000   263.750000   831.750000  1237.250000     8.000000   
50%      10.000000   564.500000  1250.000000  2153.500000    12.000000   
75%      16.000000   903.000000  1637.750000  3065.500000    16.000000   
max      20.000000  1907.000000  1998.000000  3989.000000    19.000000   

              sc_w    talk_time      three_g  touch_screen         wifi  
count  1000.000000  1000.000000  1000.000000    1000.00000  1000.000000  
mean      5.316000    11.085000     0.756000       0.50000     0.507000  
std       4.240062     5.497636     0.429708       0.50025     0.500201  
min       0.000000     2.000000     0.000000       0.00000     0.000000  
25%       2.000000     6.750000     1.000000       0.00000     0.000000  
50%       5.000000    11.000000     1.000000       0.50000     1.000000  
75%       8.000000    16.000000     1.000000       1.00000     1.000000  
max      18.000000    20.000000     1.000000       1.00000     1.000000  

[8 rows x 21 columns]

**DataFrame.describe(percentiles=None, include=None, exclude=None)**
* percentiles normal (25%, 59%, 75% (default)
* include only numeric values (default)
* exclude NaN values (default)


In [24]:
#"None" is the default...means consider only numerical values
dataset.describe(percentiles=None, exclude=None)

id  battery_power         blue  clock_speed     dual_sim  \
count  1000.000000    1000.000000  1000.000000  1000.000000  1000.000000   
mean    500.500000    1248.510000     0.516000     1.540900     0.517000   
std     288.819436     432.458227     0.499994     0.829268     0.499961   
min       1.000000     500.000000     0.000000     0.500000     0.000000   
25%     250.750000     895.000000     0.000000     0.700000     0.000000   
50%     500.500000    1246.500000     1.000000     1.500000     1.000000   
75%     750.250000    1629.250000     1.000000     2.300000     1.000000   
max    1000.000000    1999.000000     1.000000     3.000000     1.000000   

                fc       four_g   int_memory        m_dep   mobile_wt  ...  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.00000  ...   
mean      4.593000     0.487000    33.652000     0.517500   139.51100  ...   
std       4.463325     0.500081    18.128694     0.280861    34.85155  ...   
min       0.000000     0.000000     2.000000     0.100000    80.00000  ...   
25%       1.000000     0.000000    18.000000     0.300000   109.75000  ...   
50%       3.000000     0.000000    34.500000     0.500000   139.00000  ...   
75%       7.000000     1.000000    49.000000     0.800000   170.00000  ...   
max      19.000000     1.000000    64.000000     1.000000   200.00000  ...   

                pc    px_height     px_width          ram         sc_h  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     10.054000   627.121000  1239.774000  2138.998000    11.995000   
std       6.095099   432.929699   439.670981  1088.092278     4.320607   
min       0.000000     0.000000   501.000000   263.000000     5.000000   
25%       5.000000   263.750000   831.750000  1237.250000     8.000000   
50%      10.000000   564.500000  1250.000000  2153.500000    12.000000   
75%      16.000000   903.000000  1637.750000  3065.500000    16.000000   
max      20.000000  1907.000000  1998.000000  3989.000000    19.000000   

              sc_w    talk_time      three_g  touch_screen         wifi  
count  1000.000000  1000.000000  1000.000000    1000.00000  1000.000000  
mean      5.316000    11.085000     0.756000       0.50000     0.507000  
std       4.240062     5.497636     0.429708       0.50025     0.500201  
min       0.000000     2.000000     0.000000       0.00000     0.000000  
25%       2.000000     6.750000     1.000000       0.00000     0.000000  
50%       5.000000    11.000000     1.000000       0.50000     1.000000  
75%       8.000000    16.000000     1.000000       1.00000     1.000000  
max      18.000000    20.000000     1.000000       1.00000     1.000000  

[8 rows x 21 columns]

Considering the percentiles of *battery_power* column:

*   (25%) - 25% of the records have the value 895 or below
*   (50%) - 50% of the records have the value 1246.5 or below
*   (75%) - 75% of your records have the value 1629.25 or below

In [ ]:
dataset.describe(include='all')



---


Practice: Explore the simple Lab

In [25]:
import pandas as pd

data = {'A': [1, 2, 3, 4, 5], 'B': [np.nan, 20, 30, 40, 50], 'C': ['a', 'b', 'c', 'd', 'c']}
df = pd.DataFrame(data)

In [26]:
df

A     B  C
0  1   NaN  a
1  2  20.0  b
2  3  30.0  c
3  4  40.0  d
4  5  50.0  c

In [27]:
df.shape

(5, 3)

In [ ]:
#basic statistics for numerical only
print(df.describe())

In [ ]:
df.describe(include='all')



---



In [ ]:
#What it does?
df.describe(percentiles=None, exclude=None)

#and this
df.describe(percentiles=[0.1, 0.9])

**Statistical calculus for a particular column**

*Count number of non-NA/null observations.*

In [28]:
dataset['battery_power'].count()

np.int64(1000)

*Maximum of the battery_power values*

In [29]:
dataset['battery_power'].max()

1999

*Mean of the battery_power values*

In [ ]:
dataset['battery_power'].mean()

*Minimum value of battery_power*

In [ ]:
dataset['battery_power'].min()

*Median of battery_power column*

In [ ]:
dataset['battery_power'].median()

*Check how may occurences for each value of a particulaar column*

In [ ]:
dataset['battery_power'].value_counts()

*Select part of the columns*

In [ ]:
dataset[['battery_power','dual_sim']]

*Details about part of the columns*

In [ ]:
dataset[['battery_power','dual_sim']].describe(include="all")

*Details about a partivular column*

In [ ]:
dataset.battery_power.describe()

### Considering columns of type *objet* (text)

By default, only numeric values are considered. To analyse objects columns (text values) use "*dataset.describe(include=object)*"

*Example of Dataframe with objects:*

In [ ]:
df = pd.DataFrame({'gender': pd.Categorical(['Male', 'Female', 'Others']),
                   'quantity': [12, 25, 3],
                   'symbol': ['M', 'F', 'O']
                   })
df

*By default omly numerical values are described*

In [ ]:
df.describe()

*Describing all columns of a DataFrame regardless of data type*

In [ ]:
df.describe(include="all")

*Including only string columns in the DataFrame description.*

In [ ]:
df.describe(include=object)
#or
df.describe(include=[object])

*Excluding numeric columns from a DataFrame description.*

In [ ]:
df.describe(exclude=[np.number])

*Including only numeric columns in the DataFrame description.*

In [ ]:
df.describe(include=[np.number])

*Only Categorical columns*

In [ ]:

df.describe(include=['category'])

*Excluding object columns from a DataFrame description.*

In [ ]:
df.describe(exclude=[object])

### Filtering Data

The idea it to *get filtered data* for any criterium.

Column value greater than...

In [ ]:
#Back to original dataset
newdf = dataset.copy()
newdf = dataset.query('battery_power>1000')
newdf

Filtering one column and compare with another

In [ ]:
newdf.loc[newdf["battery_power"] >1500, "clock_speed"].value_counts()

Complex filtering

In [ ]:
newdf[(newdf["clock_speed"] < 1) & (newdf["dual_sim"] ==1)].groupby(\

    ["battery_power", "mobile_wt","clock_speed"])["int_memory"].count()

*Calculus over filtered data*

In [ ]:
dataset.groupby('dual_sim').sum()

In [ ]:
dataset.groupby('dual_sim').count()

In [ ]:
dataset.groupby("dual_sim", sort=False)["n_cores"].sum()

## 4- Final Pratice

In [ ]:
#explore the dataframe
data = {
    'A': ['apple', 'banana', 'apple', 'orange', 'banana'],
    'B': [10, 20, None, 40, 50],
    'C': ['x', 'y', 'x', 'z', None]
}
df = pd.DataFrame(data)

## References


*   Using pandas and Python to Explore Your Dataset
https://realpython.com/pandas-python-explore-dataset/




End!